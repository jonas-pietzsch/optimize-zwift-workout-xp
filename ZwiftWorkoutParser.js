import { XMLParser, XMLBuilder } from "fast-xml-parser";
import { ZwiftWorkout } from "./ZwiftWorkout.js";

const options = {
  ignoreAttributes: false,
  attributeNamePrefix: "attribute",
  preserveOrder: true,
};

export class ZwiftWorkoutParser {
  static parseZwiftWorkoutFile(fileBuffer) {
    const parser = new XMLParser(options);
    const parsedWorkoutFile = parser.parse(fileBuffer.toString());
    return new ZwiftWorkout(parsedWorkoutFile);
  }

  static assembleZwiftWorkoutFile(zwiftWorkout) {
    const builder = new XMLBuilder({
      ...options,
      format: true,
    });
    return builder.build(zwiftWorkout.contents);
  }
}
