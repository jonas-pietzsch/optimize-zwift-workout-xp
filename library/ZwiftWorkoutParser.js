import { XMLParser, XMLBuilder } from "fast-xml-parser";
import { ZwiftWorkout } from "./ZwiftWorkout.js";

const options = {
  ignoreAttributes: false,
  attributeNamePrefix: "attribute",
  preserveOrder: true,
};

export class ZwiftWorkoutParser {
  static xmlParser = new XMLParser(options);
  static xmlBuilder = new XMLBuilder({ ...options, format: true });

  static parseZwoFile(fileBuffer) {
    const parsedWorkoutFile = ZwiftWorkoutParser.xmlParser.parse(
      fileBuffer.toString()
    );
    return new ZwiftWorkout(parsedWorkoutFile);
  }

  static assembleZwoFile(zwiftWorkout) {
    return ZwiftWorkoutParser.xmlBuilder.build(zwiftWorkout.contents);
  }
}
