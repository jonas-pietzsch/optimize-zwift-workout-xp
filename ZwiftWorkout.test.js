import { ZwiftWorkoutParser } from "./ZwiftWorkoutParser";
import * as fs from "fs";

describe("ZwiftWorkout", () => {
  describe("optimize", () => {
    it("should optimize the workout", () => {
      const sampleWorkout = ZwiftWorkoutParser.parseZwiftWorkoutFile(
        fs.readFileSync("./sample.zwo")
      );
      sampleWorkout.optimize({
        minimumDuration: 60,
        intervalsDuration: 30,
      });

      expect(sampleWorkout.contents).toMatchSnapshot();
    });
  });
});
