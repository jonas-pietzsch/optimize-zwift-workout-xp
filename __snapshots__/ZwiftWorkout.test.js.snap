// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZwiftWorkout optimize should optimize the workout 1`] = `
[
  {
    "workout_file": [
      {
        "author": [
          {
            "#text": "Somebody",
          },
        ],
      },
      {
        "name": [
          {
            "#text": "Something",
          },
        ],
      },
      {
        "description": [
          {
            "#text": "Some description",
          },
        ],
      },
      {
        "sportType": [
          {
            "#text": "bike",
          },
        ],
      },
      {
        "tags": [],
      },
      {
        "workout": [
          {
            ":@": {
              "attributeDuration": "900",
              "attributePowerHigh": "0.65",
              "attributePowerLow": "0.45",
            },
            "Warmup": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "15",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "15",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeDuration": "1080",
              "attributePowerHigh": "0.43",
              "attributePowerLow": "0.63",
            },
            "Cooldown": [],
          },
        ],
      },
    ],
  },
]
`;
